File: apis/users-service/src/handlers/ErrorHandler/ErrorHandler.ts
import IErrorHandler from './interface';
import BaseError from '@constants/BaseError';
import ILoggerProvider from '@providers/LoggerProvider/interface';

export default class ErrorHandler implements IErrorHandler {
  private loggerProvider: ILoggerProvider;

  constructor(loggerProvider: ILoggerProvider) {
    this.loggerProvider = loggerProvider;
  }

  public async handleError(error: Error): Promise<void> {
    if (this.isTrustedError(error)) {
      this.loggerProvider.error('', error);
    } else {
      this.loggerProvider.fatal('', error);
    }
  }

  public isTrustedError(error: Error): boolean {
    if (error instanceof BaseError) {
      return error.isOperational;
    }

    return false;
  }
}

File: apis/users-musics-service/internal/handlers/errorHandler/errorHandler.go
package errorHandler

import (
	"users-musics-service/internal/constants"
	l "users-musics-service/internal/providers/loggerProvider"
)

type errorHandler struct {
	loggerProvider l.ILoggerProvider
}

func New(loggerProvider l.ILoggerProvider) IErrorHandler {
	return errorHandler{
		loggerProvider: loggerProvider,
	}
}

func (e errorHandler) HandleError(err error) {
	if e.IsTrustedError(err) {
		e.loggerProvider.Error("", err)
	} else {
		e.loggerProvider.Fatal("", err)
	}
}

func (e errorHandler) IsTrustedError(err error) bool {
	if customError, ok := err.(*constants.BaseError); ok {
		return customError.IsOperational
	}

	return false
}
